usr/lib/api/api_interface.c

```diff
@@ -1705,14 +1706,27 @@ CK_RV C_Finalize(CK_VOID_PTR pReserved)
     bt_destroy(&Anchor->sess_btree);
 
 #if OPENSSL_VERSION_PREREQ(3, 0)
-    ERR_set_mark();
-    if (Anchor->openssl_default_provider != NULL)
-        OSSL_PROVIDER_unload(Anchor->openssl_default_provider);
-    if (Anchor->openssl_legacy_provider != NULL)
-        OSSL_PROVIDER_unload(Anchor->openssl_legacy_provider);
-    if (Anchor->openssl_libctx != NULL)
-        OSSL_LIB_CTX_free(Anchor->openssl_libctx);
-    ERR_pop_to_mark();
+    /*
+     * Only cleanup OpenSSL library context and providers if we are not in the
+     * library destructor. The library destructor calls C_Finalize if not
+     * already finalized, but this may happen during at-exit handlers when the
+     * program is terminating. At that point in time, the OpenSSL at-exit
+     * handler may already have performed cleanup which will then cause
+     * crashes when trying to cleanup the already freed library context here.
+     * We are leaking the library context and providers if one just unloads
+     * the library without calling C_Finalize. However, OpenSSL cleanup will
+     * clean up the context at program termination anyway.
```

Comment:
Well, at least the loaded providers are cleaned within CONF_modules_finish() which is called in OPENSSL_cleanup() as far as I can tell. But yes, the library context may be leaked though.... Nevertheless, cleaning the context in openCryptoki may cause crashed because the providers used in that context have already been cleaned.